### ISSUE TITLE
Problems with Regularization Methods

### ISSUE THREAD

**Issue by ConteMax66** [Artelnics/opennn#241]
I've begun to work with your Library since version 5.0.5  Now passing to your new version 6.0, but the Regularization methods seems not working anymore. If you test any std network (Approximation, with any learning method) with constant seed initialization parameters, and sequential splitting training and selection datasets, you will see that the network behavement in terms of net convergence results will be exactly the same, regardless what regularization methods you choose or regularization weight, or without any regularization. I've seen you've done deep changes in this part of code inside classes and methods in files "loss_index". 
In my opinion seems that the Regularization error gradient is not propagated to the network. May be you should have a look.

**Comment by ConteMax66** [comment]
It's not a performance issue, it's more a bug in the new version of Regularization coding... IMHO.

**Comment by ConteMax66** [comment]
I found the bug in your code in file loss_index.cpp inside the function "back_propagate",
the function call "assemble_layers_error_gradient(back_propagation);" should be put after the 
call to "add_regularization_gradient"  and not before to propagate the error gradient to the "backpropagation.gradient" component that is used after for training: Here below my solution, now seems everithing fine with regularization working.
Let me know if this solution is correct.

    void LossIndex::back_propagate(const DataSetBatch& batch,
        NeuralNetworkForwardPropagation& forward_propagation,
        LossIndexBackPropagation& back_propagation) const
    {
        // Loss index

        calculate_errors(batch, forward_propagation, back_propagation);

        calculate_error(batch, forward_propagation, back_propagation);

        calculate_layers_delta(batch, forward_propagation, back_propagation);

        calculate_layers_error_gradient(batch, forward_propagation, back_propagation);


        // Loss
                       
        back_propagation.loss = back_propagation.error;

        // Regularization

        if (regularization_method != RegularizationMethod::NoRegularization)
        {
            add_regularization(back_propagation);
            add_regularization_gradient(back_propagation);
        }

        if (back_propagation.assemble)
            assemble_layers_error_gradient(back_propagation); // this call was before the "if (regularization_method...)..." statement 


    }

**Comment by davidge807** [comment]
Hi @ConteMax66 

I've been checking your solution and it's correct. I'm adding it in the next commit.

Thank you for your comment and interest in OpenNN library!
